alibi_detect/cd/tensorflow/classifier.py

```diff
@@ -176,8 +176,13 @@ def score(self, x: np.ndarray) -> Tuple[float, float, np.ndarray, np.ndarray,  #
             else:
                 raise TypeError(f'x needs to be of type np.ndarray or list and not {type(x)}.')
             ds_tr = self.dataset(x_tr, y_tr)
-            self.model = clone_model(self.original_model) if self.retrain_from_scratch \
-                else self.model
+            if self.retrain_from_scratch:
+                # clone model to re-initialise
+                self.model = clone_model(self.original_model)
+                # If a new (tf>=2.11) optimizer, clone this too to prevent error due to cloned model
+                optimizer = self.train_kwargs['optimizer']
+                if isinstance(optimizer, tf.keras.optimizers.Optimizer):  # If not a legacy optimizer
+                    self.train_kwargs['optimizer'] = optimizer.__class__.from_config(optimizer.get_config())
```

Comment:
This is "option 2" of the possible fixes. Cloning the `optimizer` as well as the `model` prevents it from complaining about the `model` changing on successive `predict` calls.